<B>PRIMER PRACTICO

1.Importar librerias y abrir dataset

In [22]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [23]:
dataSet=pd.read_csv("permisos_construccion_2.csv",low_memory=False)

2. Interpretacion

Numero de permiso:
Numero asignado al completar el formulario

Tipo de permiso:
Tipo de representado representado numericamente

Definición de Tipo de Permiso:
Descripción del tipo de permiso, por ejemplo nueva construcción, modificaciones

Fecha de Creación del Permiso:	
Fecha en que se creó el permiso, posterior o igual a la fecha de presentación

Bloque:
Relacionado con la dirección

Lote:
Relacionado con la dirección

Número de Calle:
Relacionado con la dirección

Sufijo de Número de Calle:
Relacionado con la dirección

Nombre de la Calle:
Relacionado con la dirección

Sufijo de Nombre de la Calle:
Relacionado con la dirección

Unidad:
Unidad de un edificio

Sufijo de Unidad:
Sufijo, si hay alguno, para la unidad

Descripción:
Detalles sobre el propósito del permiso. Ejemplo: retejado, renovación de baño

Estado Actual:
Estado actual de la solicitud de permiso

Fecha de Estado Actual:
Fecha en que se ingresó el estado actual

Fecha de Presentación:
Fecha de presentación del permiso

Fecha de Emisión:
Fecha de emisión del permiso

Fecha de Finalización:
Fecha en que se completó el proyecto, aplicable si Estado Actual = "completado"

Fecha del Primer Documento de Construcción:
Fecha en que se documentó la construcción

Notificación Estructural:
Notificación para cumplir con alguna necesidad legal, otorgada o no

Número de Pisos Existentes:
Número de pisos existentes en el edificio. No aplicable para ciertos tipos de permisos

Número de Pisos Propuestos:
Número de pisos propuestos para la construcción/modificación

Reforzamiento Voluntario de Pisos Blandos:
Reforzamiento de piso blando para cumplir con regulaciones sísmicas

Permiso Solo de Incendios:
Permiso relacionado con la prevención de riesgos de incendio

Fecha de Vencimiento del Permiso:
Fecha de vencimiento relacionada con el permiso emitido

Costo Estimado:
Estimación inicial del costo del proyecto

Costo Revisado:
Estimación revisada del costo del proyecto

Uso Existente:
Uso actual del edificio

Unidades Existentes:
Número actual de unidades

Uso Propuesto:
Uso propuesto del edificio

Unidades Propuestas:
Número propuesto de unidades

Juegos de Planos:
Representación del plan que indica el diseño general de la fundación

Cumplimiento TIDF:
Cumplimiento con TIDF o no, este es un nuevo requisito legal

Tipo de Construcción Existente:
Tipo de construcción existente, categorizado numéricamente

Descripción del Tipo de Construcción Existente:
Descripción del tipo anterior, por ejemplo, madera u otros tipos de construcción

Tipo de Construcción Propuesto:
Tipo de construcción propuesto, categorizado numéricamente

Descripción del Tipo de Construcción Propuesto:
Descripción del tipo anterior

Permiso de Sitio:
Permiso para sitio

Distrito del Supervisor:
Distrito del supervisor al que pertenece la ubicación del edificio

Barrios - Límites de Análisis:
Barrio al que pertenece la ubicación del edificio

Código Postal:
Código postal de la dirección del edificio

Ubicación:
Ubicación en par de coordenadas de latitud y longitud

ID de Registro:
Alguna identificación, no útil para esto

3. Analisis Exploratorio

a) Tamaño

Utilizaremos la funcion shape para poder vizualizar el tamaño del dataset, la primera coordenada son la cantidad de filas y la segunda la cantidad de columnas

In [4]:
dataSet.shape

(198910, 43)

Como se puede observar, tenemos 198910 filas y 43 columnas

b)  Mostrar primeras filas

In [48]:
dataSet.head()

Permit Number  Permit Type  Permit Type Definition Permit Creation Date  \
0       M394467            8  otc alterations permit            5/15/2013   
1   2.01505E+11            8  otc alterations permit            5/22/2015   
2       M798247            8  otc alterations permit            6/16/2017   
3   2.01308E+11            8  otc alterations permit            8/16/2013   
4   2.01307E+11            8  otc alterations permit            7/26/2013   

  Block  Lot  Street Number Street Number Suffix Street Name Street Suffix  \
0  3751  172            300                    0        03rD            St   
1  3556    6            328                    A    gUeRrErO            St   
2   655   61           2426                    A        pInE            St   
3  4267   29           2607                    A      bRyAnT            St   
4  1518   37            535                    A        26tH            Av   

   ...  Existing Construction Type Existing Construction Type Description  \
0  ...                         NaN                                    NaN   
1  ...                         5.0                         wood frame (5)   
2  ...                         NaN                                    NaN   
3  ...                         5.0                         wood frame (5)   
4  ...                         5.0                         wood frame (5)   

  Proposed Construction Type Proposed Construction Type Description  \
0                        NaN                                    NaN   
1                        5.0                         wood frame (5)   
2                        NaN                                    NaN   
3                        5.0                         wood frame (5)   
4                        5.0                         wood frame (5)   

  Site Permit Supervisor District Neighborhoods - Analysis Boundaries  \
0         NaN                 NaN                                 NaN   
1         NaN                   8                             Mission   
2         NaN                   2                     Pacific Heights   
3         NaN                   9                             Mission   
4         NaN                   1                      Outer Richmond   

   Zipcode                                   Location     Record ID  
0      NaN                                        NaN  1.304810e+12  
1  94103.0  (37.765906581390105, -122.42453043666066)  1.382530e+12  
2  94115.0   (37.78788490521045, -122.43641930200963)  1.467270e+12  
3  94110.0   (37.75241643936911, -122.40877419451499)  1.314500e+12  
4  94121.0   (37.77928233382438, -122.48608415176169)  1.312290e+12  

[5 rows x 43 columns]

c) Datos faltantes por fila

In [6]:
dataSet.isnull().sum()*100/len(dataSet)

Permit Number                              0.000000
Permit Type                                0.000000
Permit Type Definition                     0.000000
Permit Creation Date                       0.000000
Block                                      0.000000
Lot                                        0.000000
Street Number                              0.000000
Street Number Suffix                      98.885928
Street Name                                0.000000
Street Suffix                              1.391584
Unit                                      85.179227
Unit Suffix                               99.014127
Description                                0.145795
Current Status                             0.000000
Current Status Date                        0.000000
Filed Date                                 0.000000
Issued Date                                7.511940
Completed Date                            51.136192
First Construction Document Date           7.514957
Structural N

Los primeros datos que vemos con varios nulos son la unidad, el sufijo de la unidad y el sufijo del numero de calle (Street Number Suffix, Unit, Unit Suffix). Esta informacion puede no corresponder a ciertas ubicaciones y por eso es que hay varios valores nulos. Ademas, puede que hayan puesto esta informacion en la direccion directamente, sin dividir por categoria o tambien puede ser que no sepan esta informacion especifica al rellenar el formulario.

Luego, la fecha de finalizacion tiene la mitad con nulos, esto problablemente se deba a que los trabajos no han sido completados. Vemos que si hacemos la comparacion del estado con la fecha de completado, los que tienen fecha la gran mayoria se encuentran completados y los que no tienen fecha ninguno de ellos esta completo.

In [7]:
estado_sin_fecha = dataSet[dataSet['Completed Date'].isna()]['Current Status'].value_counts()
estado_con_fecha = dataSet[dataSet['Completed Date'].notna()]['Current Status'].value_counts()

In [8]:
print(estado_sin_fecha)

issued         83510
filed          12045
withdrawn       1754
cancelled       1531
expired         1363
approved         728
reinstated       527
suspend          190
revoked           46
plancheck         16
disapproved        2
appeal             2
incomplete         1
Name: Current Status, dtype: int64


In [9]:
print(estado_con_fecha)

complete      97081
issued           53
reinstated       36
expired           7
cancelled         5
approved          5
revoked           4
suspend           3
incomplete        1
Name: Current Status, dtype: int64



En el caso de Notificacion estructural, reforzmiento voluntario de pisos blandos, permiso de incendios, cumplimiento TIDF y permiso de sitio (Structural Notification, Voluntary Soft-Story Retrofit, Fire Only Permit, TIDF Compliance, Site Permit) al ser variables booleanas, el no se deja como vacio y por eso hay tantos null.

In [10]:
dataSet['Structural Notification'].value_counts()

Y    6922
Name: Structural Notification, dtype: int64

e) 

Variables Discretas:
- `Permit Type`
- `Permit Type Definition`
-   Current Status`
- SstructuralNnotification`
- `Number of Existing Stories` &#x2717;
- `Number of proposed stories` &#x2717;

- `Voluntary Soft-Story Retrfit`

- `Fire Only Permit`

- `Existing Use` 

- existing units&#x2717; 

- proposed units&#x2717; 

- tidf compliance
 
- existing construction type

- proposed construction type

- site permit

Las 2 primeras variables discretas son el tipo de permiso y su descripcion

In [11]:
dataSet['Permit Type'].value_counts()

8    178852
3     14664
4      2892
2       950
6       600
7       511
1       350
5        91
Name: Permit Type, dtype: int64

In [20]:
dataSet['Permit Type Definition'].value_counts()

otc alterations permit                 178836
additions alterations or repairs        14663
sign - erect                             2892
new construction wood frame               950
demolitions                               600
wall or painted sign                      511
new construction                          347
grade or quarry or fill or excavate        91
 otc alterations permit                     8
otc alterations permit #                    8
new construction #                          2
 new construction                           1
 additions alterations or repairs           1
Name: Permit Type Definition, dtype: int64

Como se puede observar, el tipo de permiso tiene valores que van del 1 al 8. La descripcion del tipo de permiso se deberia de corresponder con algun numero, pero se nota que hay mas definiciones. Esto se debe a que las definiciones, a pesar de ser iguales tienen algun caracter de diferencia por lo que las pone en lugares distintos.

Si agrupamos los numeros y que definicion tiene cada uno, se observa lo siguiente:

In [55]:
grupos=dataSet.groupby('Permit Type')['Permit Type Definition'].unique()
for i in range(1,9):
    print(f"Numero {i}")
    print(grupos[i])

Numero 1
['new construction' ' new construction ' 'new construction #']
Numero 2
['new construction wood frame']
Numero 3
['additions alterations or repairs' ' additions alterations or repairs ']
Numero 4
['sign - erect']
Numero 5
['grade or quarry or fill or excavate']
Numero 6
['demolitions']
Numero 7
['wall or painted sign']
Numero 8
['otc alterations permit' ' otc alterations permit '
 'otc alterations permit #']


Habria que modificar las definiciones de permiso 1, 3 y 8 para que concuerden

In [85]:
dataSet['Current Status'].value_counts()

complete       97081
issued         83563
filed          12045
withdrawn       1754
cancelled       1536
expired         1370
approved         733
reinstated       563
suspend          193
revoked           50
plancheck         16
incomplete         2
disapproved        2
appeal             2
Name: Current Status, dtype: int64

In [14]:
dataSet['Number of Existing Stories'].unique()

array([ nan,  3. ,  2. ,  1. ,  4. ,  5. , 28. ,  0. , 11. ,  6. , 40. ,
       29. ,  8. ,  9. , 18. , 30. , 24. , 23. , 12. , 38. , 42. , 20. ,
       25. ,  7. , 52. , 10. , 43. , 16. , 14. , 48. , 21. , 33. , 22. ,
       27. , 37. , 13. , 32. , 31. , 15. , 39. , 58. , 19. , 17. , 26. ,
       63. , 34. , 36. , 35. , 50. , 41. , 44. , 45. , 53. , 47. , 49. ,
       60. , 46. , 61. , 55. , 62. , 54. ,  1.5, 56. , 78. ,  2.5])

In [47]:
dataSet['Number of Existing Stories'].value_counts()

2.0     52770
3.0     45741
4.0     16055
1.0      8793
5.0      3767
        ...  
49.0        3
62.0        2
2.5         2
1.5         1
78.0        1
Name: Number of Existing Stories, Length: 64, dtype: int64

In [44]:
dataSet['Number of Proposed Stories'].unique()

array([ nan,  3. ,  2. ,  1. ,  4. ,  5. , 28. , 11. ,  0. ,  6. , 40. ,
       29. ,  8. , 21. ,  9. , 18. , 30. , 24. , 23. , 12. , 38. , 42. ,
       20. , 25. ,  7. , 52. , 10. , 43. , 16. , 14. , 48. , 33. , 63. ,
       22. , 27. , 19. , 37. , 13. , 32. , 31. , 15. , 39. , 58. , 17. ,
       26. , 34. , 36. , 35. , 50. , 41. , 44. , 45. , 53. , 46. , 47. ,
       55. , 60. , 49. , 56. , 61. , 62. , 54. ,  1.5, 78. ,  2.5])

dataSet['Number of Proposed Stories'].value_counts()

In [64]:
dataSet['Number of Proposed Stories'].value_counts()

2.0     50994
3.0     46803
4.0     17594
1.0      7893
5.0      4044
        ...  
47.0        3
49.0        3
2.5         2
1.5         1
78.0        1
Name: Number of Proposed Stories, Length: 64, dtype: int64

In [80]:
dataSet['Structural Notification'].value_counts()

Y    6922
Name: Structural Notification, dtype: int64

In [72]:
dataSet['Voluntary Soft-Story Retrofit'].value_counts()

Y    35
Name: Voluntary Soft-Story Retrofit, dtype: int64

In [82]:
dataSet['Fire Only Permit'].unique()

array([nan, 'Y'], dtype=object)

In [70]:
dataSet['Existing Use'].unique()

array([nan, 'apartments', '2 family dwelling', '1 family dwelling',
       'retail sales', 'food/beverage hndlng', 'warehouse,no frnitur',
       'office', 'massage parlor', 'barber/beauty salon', 'vacant lot',
       'storage shed', 'fence/retaining wall', 'day care home gt 12',
       'workshop commercial', 'accessory cottage', 'artist live/work',
       'residential hotel', 'auto repairs', 'prkng garage/public',
       'lending institution', 'parking lot', 'prkng garage/private',
       'stadium', 'filling/service stn', 'nursing home non amb',
       'theater', 'school', 'tourist hotel/motel', 'church', 'mortuary',
       'clinics-medic/dental', 'manufacturing', 'animal sale or care',
       'misc group residns.', 'health studios & gym',
       'sfpd or sffd station', 'christmas tree lot',
       'warehouse, furniture', 'dry cleaners', 'club', 'wholesale sales',
       'recreation bldg', 'nite club', 'power plant', 'garment shops',
       'ambulance service', 'antenna', 'social care

In [60]:
dataSet['Existing Use'].value_counts()

1 family dwelling       46768
apartments              40800
office                  24617
2 family dwelling       20988
retail sales             6910
                        ...  
bath house                  2
dairies/dairy equip.        1
child care                  1
orphanage                   1
roofing materials           1
Name: Existing Use, Length: 93, dtype: int64